### ISSUE TITLE
Could not decode media stream on neon.data.dataloader.DataLoader

### ISSUE THREAD

**Issue by smly** [anlthms/sp-2016#4]
Hi Anil,

In my env, your code works in many cases, but I got an runtime error while processing an archive file (cpio) including some specific wave data. Here is the execution log ([complete log here](https://gist.github.com/smly/a0cf1506db570b0085487b3d76c6dc12/raw/e6226060a628606e6ead8ff7b614271524dedf65/crash_log.log)).: 

```
(snip)
  Linear Layer 'Linear_0': 256 inputs, 2 outputs
  Activation Layer 'Linear_0_Softmax': Softmax
terminate called after throwing an instance of 'std::runtime_error'
  what():  Could not decode media stream
./run.sh: line 45: 11211 Aborted                 (core dumped) ./model.py -e $num_epochs -w $train_dir -r 0 -z 64 -v --no_progress_bar -elec $elec -out $out_dir
```

By debugging, I found this exception is raised on loader/src/codec.hpp:L139. The returned value from avcodec_decode_audio4() is -1094995529. This error code indicates AVERROR_INVALIDDATA (-0x41444E49, INDA). On my environment, it is [reproducible when processing 2_1418_0.3.wav](https://gist.github.com/smly/a0cf1506db570b0085487b3d76c6dc12#file-test_dataloader_audioparams-py-L23). Here is the test code to reproduce the runtime error:
https://gist.github.com/smly/a0cf1506db570b0085487b3d76c6dc12#file-test_dataloader_audioparams-py

For now, I'm not sure the detail but I think this error causes with using recent version of libavcodec. Here is the version information on my environment ([detail](https://gist.github.com/smly/a0cf1506db570b0085487b3d76c6dc12#file-version)).

```
* Ubuntu 16.04
* libavutil 54, libavcodec 56, libavformat 56
```

Do you have any information about this error? (or should I report this issue on NervanaSystems/neon?) Anyway, thank you very much for sharing this interesting example on processing signal data.

Kohei

**Comment by anlthms** [comment]
Hi Kohei,

Thanks for reporting this. As you implied, this seems to be an issue with libav. I wonder if you will run into the same error if you try to play the wave file in question with ffmpeg.

Just to be sure, you might want to try recreating the archive (cpio) files. They get generated automatically if you delete them using clear.sh.

You might also want to try unpacking the archive file using the cpio command and making sure that the unpacked wave files are exactly the same as the originals.

There's also the possibility that scikits.audiolab is creating corrupt wave files. You could try using some other library (e.g. scipy.io.wavfile) to create the wave files in that case.

Anil

**Comment by ternaus** [comment]
@smly 

I get the same error. Any success?

P.S. Last commit for AudioLab was 4 years ago.

**Comment by RishabGargeya** [comment]
@smly @ternaus Ran into this same issue. Any luck finding a workaround? I'm also on 16.04 with the same environment version information. I tried replacing audiolab with scipy.io.wavefile's write function, but it didn't solve the issue. Thanks!

**Comment by ternaus** [comment]
Partially helped to filter frequencies to [0.1 - 180Hz]. 

Still not everything is working. But problem is less severe, meaning less broken files.

```
#!/usr/bin/env python
"""
Extract the data out of .mat files and save as .wav files.
"""

import sys
import os
import glob
import numpy as np
from scipy import io
from scipy.io.wavfile import write
from scipy.signal import spectrogram
from scipy.signal import butter, lfilter

def extract(path):
    print('Extracting data into %s...' % path)
    filelist = glob.glob(os.path.join(path, '*.mat'))
    for srcfile in filelist:
        wavwrite(srcfile)


def butter_bandpass(lowcut, highcut, fs, order=5):
    nyq = 0.5 * fs
    low = lowcut / nyq
    high = highcut / nyq
    b, a = butter(order, [low, high], btype='band')
    return b, a


def butter_bandpass_filter(data, lowcut, highcut, fs, order=5):
    b, a = butter_bandpass(lowcut, highcut, fs, order=order)
    y = lfilter(b, a, data)
    return y


def f(x):
    return butter_bandpass_filter(x, 0.1, 180, 400, order=5)


def wavwrite(srcfile):
    try:
        mat = io.loadmat(srcfile, verify_compressed_data_integrity=False)
    except ValueError:
        print('Could not load %s' % srcfile)
        return

    dat = mat['dataStruct'][0, 0][0]
    mn = dat.min()
    mx = dat.max()
    mx = float(max(abs(mx), abs(mn)))
    if mx != 0:
        dat *= 0x7FFF / mx

    for elec in range(16):
        dstfile = srcfile.replace('mat', str(elec) + '.wav')
        aud = dat[:, elec] - np.mean(dat[:, elec])

        write(dstfile, 400, np.int16(f(aud)))


if len(sys.argv) < 2:
    print('Usage %s /path/to/data' % sys.argv[0])
    sys.exit(0)

for subj_id in range(1, 4):
    for prefix in ['train_', 'test_']:
        path = os.path.join(sys.argv[1], prefix + str(subj_id))
        extract(path)

```

**Comment by RishabGargeya** [comment]
Still unable to completely fix the issue, tried neon 1.6.0 and followed https://github.com/NervanaSystems/neon/issues/302
to install ffmpeg and remake neon, but the problem persists.

I run into the error on the following command on my setup- 

```
./model.py -e 16 -w ../input//train_2/ -r 0 -z 64 -v --no_progress_bar -elec 3 -out ../output/
```

@anlthms Do you have any other ideas on what could cause this discrepancy? Thanks so much for your help.

**Comment by anlthms** [comment]
@smly, I don't have a solution, but I can confirm that I have the same checksum for the problematic file. I tried running your tests (I hadn't noticed them before) and did not encounter the error. But I'm on ubuntu 14.04 and probably have an older version of libav than what you have.

```
testcase_elec2()
archive is successfully generated on debug1-ingested/
data_loader is ok
testcase_elec3()
archive is successfully generated on debug1-ingested/
data_loader is ok
```

At this point, it would be interesting to check if ffmpeg is able to decode the file in your environment.
